Add support for For loops to quasiquotes

densh opened the pull request:
review by @retronym @xeno-by

densh committed:
deduplicate tuple tree creation code

Previously tuple tree generation code has been implemented in
three place: tree builder, tree gen, build utils. Now it's just
defined once in tree gen.

densh committed:
make internal implementation of universe.build less restrictive

1. Use protected instead of private to avoid needless lock-in

2. Use BuildImpl field type to expose non-protected members to the
   compiler (user-facing side in the reflection api stays the same)

densh committed:
simplify imports in quasiquotes scalacheck tests

densh committed:
add support for importable attachments

Previously attachments weren't imported by importTree. Now a new marker
trait has been added that lets attachments to import themselves to the
new universe together with all their innards. Additionally a simpler
subtrait is defined to mark attachments that can be imported as-is.

densh committed:
make fresh names compare similar in quasiquotes

Otherwise it's annoying that trees that look the same are not equal
due to some fresh name hidden underneath. Due to this one test needs
to be changed to use plain equalsStructure.

densh committed:
add hasAttachment utility method to the internal api

densh committed:
add support for for loops and for enumerators to quasiquotes

1. q"for (..$enums) $body", q"for (..$enums) yield $body"

2. fq"..." quote to construct/deconstruct enumerators

densh committed:
implement inverse transformation to mkFor

This effectively reconstructs a sequence of enumerators and body
from the tree produced by mkFor. This lets to define bi-directional
SyntacticFor and SyntacticForYield constructors/extractors to work
with for loops.

Correctness of the transformation is tested by a scalacheck test
that generates a sequence of random enumerators, sugars them into
maps/flatMaps/foreach/withFilter calls and reconstructs them back.

densh committed:
add some post-typecheck tests for quasiquotes

Typecheck trees with toolbox and check that they are still matched
by corresponding quasiquote. Fix tuples and function types matchers
to account for different shape of trees after typing.

xeno-by commented on src/reflect/scala/reflect/internal/Trees.scala:
```diff
@@ -490,7 +490,7 @@ trait Trees extends api.Trees {
   case class Ident(name: Name) extends RefTree with IdentContextApi {
     def qualifier: Tree = EmptyTree
-    def isBackquoted = this.attachments.get[BackquotedIdentifierAttachment.type].isDefined
+    def isBackquoted = this.hasAttachment[BackquotedIdentifierAttachment.type]
```
+1

xeno-by commented on src/reflect/scala/reflect/internal/StdAttachments.scala:
```diff
@@ -14,6 +14,7 @@ trait StdAttachments {
     def setAttachments(attachments: scala.reflect.macros.Attachments { type Pos = Position }): this.type = { rawatt = attachments; this }
     def updateAttachment[T: ClassTag](attachment: T): this.type = { rawatt = rawatt.update(attachment); this }
```
Given that there's been a very non-obvious problem with attachments not being imported properly, I think it's necessary to require that all attachments subclass `trait Attachment { def importAttachment(importer: Importer): this.type }`. This way every attachment author will be forced about thinking how their attachment should behave wrt imports.

retronym replied:
I'd like to see someone try that for `ConstructorDefaultsAttachment` !

``` scala
class ConstructorDefaultsAttachment(val classWithDefault: ClassDef, var companionModuleClassNamer: Namer)
```

xeno-by replied:
Then it will be an UnsupportedOperationException, and we will know about this problem upfront.

densh replied:
I don't see much of a difference between conditional inheritance from `ImportableAttachment` and a method that will usually be defaulted to throwing and exception other than the fact that later breaks compatibility for no good reason.

xeno-by replied:
The former is safer, because it will require a conscious decision from the author of the attachment, so I suggest we stick to that.

densh replied:
I seriously doubt it. The only thing it will require is copy-paste after you get a type error error. 

@retronym any views on the this question?

xeno-by replied:
Well, if we already had a similar mechanism in place, then this changeset wouldn't have to happen. This already speaks in favor of it. Also we can just compare the number of attachments that can be imported as opposed to the number of those that cannot. I'm afk now though, so I can't do that now.

adriaanm replied:
Since there's no clear winner, I'm okay leaving it at the status quo, or a follow-up PR.

xeno-by commented:
Reproducing the comment here so that it doesn't get removed by the rebase.

Given that there's been a very non-obvious problem with attachments not being imported properly, I think it's necessary to require that all attachments subclass trait Attachment { def importAttachment(importer: Importer): this.type }. This way every attachment author will be forced about thinking how their attachment should behave wrt imports.

densh commented:
@xeno-by 
1. Won't it break existing code that relies on the previous unconstrained attachments api?
2. Is tree import a crucial part of reflection api that user should really know about? Shouldn't we rather look for ways to do things automatically in this area rather than exposing such implementation details to the users?

xeno-by commented:
1. It's a major release, so we have luxury of breaking stuff, especially given that attachments are an obscure API, new to 2.10, that noone except for us knows about.
2. Here it's about compiler developers, who represent 99.99% of attachment users (never heard of anyone using attachments in macros). I'd like to prevent them from making mistakes like the one that you've fixed.
3. For that matter, we should really deprecate the attachment api in scala.reflect.api and leave it for internal use. Upd. I'll submit a pull request after OOPSLA.

gkossakowski commented:
@xeno-by: attachments are the only way to access unexpanded macro call. We need an api for that.

retronym commented on src/reflect/scala/reflect/internal/StdAttachments.scala:
```diff
@@ -14,6 +14,7 @@ trait StdAttachments {
     def setAttachments(attachments: scala.reflect.macros.Attachments { type Pos = Position }): this.type = { rawatt = attachments; this }
     def updateAttachment[T: ClassTag](attachment: T): this.type = { rawatt = rawatt.update(attachment); this }
     def removeAttachment[T: ClassTag]: this.type = { rawatt = rawatt.remove[T]; this }
+    def hasAttachment[T: ClassTag]: Boolean = rawatt.get[T].nonEmpty
```
We should implement `hasAttachment` directly, rather than internal of `get`, for better performance. I saw a small improvement in some of my benchmarks when I did this a few day ago.

retronym commented on src/reflect/scala/reflect/internal/TreeGen.scala:
```diff
@@ -454,4 +466,340 @@ abstract class TreeGen extends macros.TreeBuilder {
+      atPos(rhs.pos)(Apply(Select(rhs, nme.withFilter), visitor :: Nil))
+    }
+
+  /** If tree is a variable pattern, return Some("its name and type").4
```
Stray ".4"

densh replied:
Removed.

JamesIry commented:
PLS REBUILD ALL

scala-jenkins commented:
(kitty-note-to-self: ignore 27503611)
:cat: Roger! Rebuilding pr-scala for f17abdd4, 8ef70324, 966db752, 75e7807e, ead5c4eb, ce39d2ab, 6db0e996, 58fb90af, 9f14e129, 92e122e0. :rotating_light:

densh committed:
change intermidiate representation of for loop enumerators

Encode values into real trees rather than non-tree case classes.
This is needed for re-usability of desugaring code between quasiquotes
and parser.

densh committed:
move for loop desugaring into tree gen

densh committed:
add syntactic combinators that represent enumerators

densh committed:
re-implement hasAttachment directly in raw attachments

retronym commented:
Would be interesting to see if the new `ForAttachment` could be used to create better diagnostics on an error messages:

Now:

```
object O { def map(a: Int) = a };
scala> for (x <- O) yield x
<console>:9: error: missing parameter type
              for (x <- O) yield x
                   ^
```

Could we add ("error in expansion of for comprehension, expanded code: O.map(x => x)")?

retronym commented:
Overall this is really impressive work. Just needs a bit more documentation to explain the tricks used to represent everything as trees, then we can merge.

adriaanm commented:
ping @densh -- the M7 deadline's approaching

densh commented:
@adriaanm fixes coming in next 24h, sorry for the delay.

densh committed:
test legacy .filter support in for loop resugaring

densh commented:
> Could we add ("error in expansion of for comprehension, expanded code: O.map(x => x)")?

Yes, this should be possible. We can discuss the details after the meetings today.

densh committed:
add comments that explain new for loop enumerator encoding

densh committed:
add comments that explain FreshName extractor from Quasiquotes cake

densh commented:
PLS REBUILD/pr-scala@4405a64

densh commented:
PLS REBUILD/pr-scala@2cd95fe

scala-jenkins commented:
(kitty-note-to-self: ignore 28349971)
:cat: Roger! Rebuilding pr-scala for 4405a641. :rotating_light:

scala-jenkins commented:
(kitty-note-to-self: ignore 28349974)
:cat: Roger! Rebuilding pr-scala for 2cd95fed. :rotating_light:

adriaanm commented:
All green! Thanks, @densh!

adriaanm commented:
LGTM proxy from @retronym, whom I just spoke to on the telephone

adriaanm commented:
+1 on the impressive work bit, also!
